### PR TITLE
Feature/logger

### PR DESCRIPTION
**Pull request by andriy-kozlytski**
https://app.asana.com/0/0/1201656496013916/f

### CODE REVIEW

File: src/Utils/LoggerHandler.php [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+		$this->filename = $filename;
+	}
+
+    public function handle(array $vars) :void
+    {
+		$output = self::DEFAULT_FORMAT;
+		foreach ($vars as $var => $value) {
+			$output = str_replace('%' . $var . '%', $value, $output);
+		}
+		file_put_contents($this->filename, $output . PHP_EOL, FILE_APPEND);
```
Comment (lodry) [inline src/Utils/LoggerHandler.php]: use async put

File: src/Services/LoggerService.php [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+{
+    public const LEVELS = [
+		'DEBUG' => ['ERROR', 'WARNING', 'INFO', 'DEBUG'],
+		'PROD' => ['ERROR', 'WARNING'],
+		'DEFAULT' => ['ERROR'],
+	];
+    private $handler;
+    private $mode = 'DEFAULT';
+
+	public function __construct()
```
Comment (lodry) [inline src/Services/LoggerService.php]: remove it

File: src/Services/LoggerService.php [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+use Langivi\ImportantReminder\Interfaces\LoggerHandlerInterface;
+
+class LoggerService
+{
+    public const LEVELS = [
+		'DEBUG' => ['ERROR', 'WARNING', 'INFO', 'DEBUG'],
+		'PROD' => ['ERROR', 'WARNING'],
+		'DEFAULT' => ['ERROR'],
+	];
+    private $handler;
```
Comment (lodry) [inline src/Services/LoggerService.php]: add types

File: src/Controllers/IndexController.php [hunk 4]
```diff
@@ -2,20 +2,22 @@
...
 
-use Langivi\ImportantReminder\Loader;
-use Langivi\ImportantReminder\Services\TestService;
+use Langivi\ImportantReminder\Services\LoggerService;
+
+
 
 class IndexController extends AbstractController
 {
     private readonly ContainerBuilder $containerBuilder;
```
Comment (lodry) [inline src/Controllers/IndexController.php]: rm this

File: src/Services/LoggerService.php (hunk 3 above)
Comment (andriy-kozlytski) [reply src/Services/LoggerService.php]: enum of types?  
or whart?

File: src/Utils/LoggerHandler.php [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+namespace Langivi\ImportantReminder\Utils;
+
+use Langivi\ImportantReminder\Interfaces\LoggerHandlerInterface;
+use UnexpectedValueException;
+
+class LoggerHandler implements LoggerHandlerInterface
+{
+    private $filename ='';
+
+	public function __construct()
```
Comment (lodry) [inline src/Utils/LoggerHandler.php]: remove constructor please

File: src/Utils/LoggerHandler.php [hunk 6]
```diff
@@ -0,0 +1,39 @@
+<?php 
+
+namespace Langivi\ImportantReminder\Utils;
+
+use Langivi\ImportantReminder\Interfaces\LoggerHandlerInterface;
+use UnexpectedValueException;
+
+class LoggerHandler implements LoggerHandlerInterface
+{
+    private $filename ='';
```
Comment (lodry) [inline src/Utils/LoggerHandler.php]: add type

File: src/Services/LoggerService.php (hunk 3 above)
Comment (lodry) [reply src/Services/LoggerService.php]: add type for field

File: log/2022-01-14.log [hunk 7]
```diff
@@ -0,0 +1,19 @@
+2022-01-14T16:18:13+00:00 [INFO]: get index
```
Comment (lodry) [inline log/2022-01-14.log:1]: exclude it please

File: src/Services/LoggerService.php [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+    {
+        $result = '';
+        foreach ($context as $key => $value) {
+            $result = $result . '[' . $key . ']=' . $value . '; ';
+        }
+        return $result;
+    }
+
+    public function error($message, array $context = array()) {
+        $this->log('EROOR', $message, $context);
```
Comment (lodry) [inline src/Services/LoggerService.php]: typo

File: server.php [hunk 9]
```diff
@@ -36,21 +39,32 @@ function servePublic(string $path, HttpResponse $res, finfo $fileinfo): void
      * @var $router Router
      */
     $router = $loader->getContainer()->get('router');
-    
     $metod = HttpMethods::tryFrom($req->method);
```
Comment (lodry) [inline server.php]: typo

File: src/Interfaces/LoggerHandlerInterface.php [hunk 10]
```diff
@@ -0,0 +1,9 @@
+<?php 
+
+namespace Langivi\ImportantReminder\Interfaces;
+
+interface LoggerHandlerInterface
+{
+	public const DEFAULT_FORMAT = '%timestamp% [%level%]: %message% %context%';
```
Comment (lodry) [inline src/Interfaces/LoggerHandlerInterface.php]: excessive space

File: src/Services/LoggerService.php [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+        $this->handler = $handler;
+    }
+
+    public function setMode($mode) 
+    {
+        $this->mode = $mode;
+    }
+
+
+    public function log($level, $message, array $context = array())
```
Comment (lodry) [inline src/Services/LoggerService.php]: types?

File: src/Services/LoggerService.php [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+                'context' => self::contextToString($context),
+            ]);
+        }
+    }
+
+	protected static function contextToString(array $context): string
+    {
+        $result = '';
+        foreach ($context as $key => $value) {
+            $result = $result . $key . ' = ' . $value . '; ';
```
Comment (lodry) [inline src/Services/LoggerService.php]: better `$result .=  $key` then `$result = $result . $key .`

File: src/Utils/LoggerHandler.php [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+namespace Langivi\ImportantReminder\Utils;
+
+use Langivi\ImportantReminder\Interfaces\LoggerHandlerInterface;
+use UnexpectedValueException;
+
+class LoggerHandler implements LoggerHandlerInterface
+{
+    private string $filename ='';
+
+	public function setFilename($filename)
```
Comment (lodry) [inline src/Utils/LoggerHandler.php]: better setFileName

File: src/Utils/LoggerHandler.php [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+	{
+        $dir =  dirname($filename);
+		if (!file_exists($dir)) {
```
Comment (lodry) [inline src/Utils/LoggerHandler.php]: better check on is_dir

**Comment by lodry** [comment]
Please also remove from stubs $use_include_path

**Review by lodry — APPROVED** [review]

**Pull request merged**